lib/codegen/src/simple_preopt.rs

```diff
@@ -536,11 +542,279 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
     }
 }
 
+struct BranchOptInfo {
+    br_inst: Inst,
+    cmp_arg: Value,
+    destination: Ebb,
+    args: ValueList,
+    kind: BranchOptKind,
+}
+
+enum BranchOptKind {
+    EqualZero,
+    NotEqualZero,
+}
+
+fn branch_opt(pos: &mut FuncCursor, inst: Inst) {
+    let info = match pos.func.dfg[inst] {
+        InstructionData::BranchInt {
+            opcode: Opcode::Brif,
+            cond: br_cond,
+            destination,
+            ref args,
+        } => {
+            let first_arg = {
+                let args = pos.func.dfg.inst_args(inst);
+                args[0]
+            };
+            if let ValueDef::Result(iconst_inst, _) = pos.func.dfg.value_def(first_arg) {
+                if let InstructionData::BinaryImm {
+                    opcode: Opcode::IfcmpImm,
```

Comment:
This is admittedly not document as clearly as it could be, but we currently don't expect frontends to produce flags-using instructions like `ifcmp` or `ifcmp_imm` themselves. Regular `icmp` is lowered to `ifcmp` by the legalizer on platforms that use a flags register on x86 and in the future other architectures. Since the code here is in simple_preopt, the legalizer hasn't run yet, so in theory we don't expect to see `ifcmp_imm` occur at this point. So one question here is: are you seeing `ifmp_imm` at this point?

If so, I'm not strongly attached to the "frontends shouldn't use flags" rule; we could just as easily have frontends produce flags-using code and have the legalizer lower it into other things on platforms that don't use flags, and it's not immediately obvious which way is better.

Reply:
Error on my end. I've fixed this to match the branch ops that frontends are actually supposed to use. I was blindly going by the code example in #606 and hadn't tested it against real code until today. :)